### PR TITLE
Enable SVE ACLE implementation for tanH Aten op for FP32 dType.

### PR DESCRIPTION
**Pull request by maajidkhann**
In deep learning models, the tanh (hyperbolic tangent) function is a widely used activation function, primarily in feedforward networks, recurrent neural networks (RNNs), and various other architectures.

Also, the tanh (hyperbolic tangent) function is commonly used in **Physics-Informed Neural Networks (PINNs).** PINNs are a class of machine learning models designed to solve partial differential equations (PDEs) by incorporating the governing physics directly into the loss function, along with data-driven terms.

In PINNs, activation functions like tanh are used in the neural network architecture to enable the model to learn complex mappings between inputs (such as spatial and temporal coordinates) and outputs (such as field variables).

**Operator: tanh()**
**Current Implementation in OSS in ATen Backend:**
**SVE Flow:** Uses SVE sleef when available else std implementation.

**With this PR :**
**SVE Flow:** Uses SVE ACLE implementation. (Faster Implementation)

**Here are the performance improvements.**
**Single core perf numbers:**
image

**Metric:**  CPU time avg time per iteration (In ms)

As you can see with both gcc and clang compilers, we see a significant performance gain with SVE ACLE implementation over current OSS Implementation (Sleef) and also Neon.

**Hardware:** m7g.8xlarge (Graviton 3 Instance)

**Script used in benchmarking:**
```python
import os
#os.environ["ATEN_CPU_CAPABILITY"] = "default"
os.environ["ATEN_CPU_CAPABILITY"] = "sve256"

import torch
import torch.nn as nn

#Set the random seed for reproducibility
torch.manual_seed(1)

#Create a tensor of shape (8521, 50)
x = torch.randn(8521, 50)

for i in range(10):
output = x.tanh()

#Perform the tanh operation 1000 times and profile the performance
print("### CPU tanh")
with torch.autograd.profiler.profile(record_shapes=True) as prof:
for i in range(1000):
output = x.tanh()

#Print the profiling results sorted by self CPU time
print(prof.key_averages().table(sort_by="self_cpu_time_total"))

#Optionally print the final output (if needed, uncomment the following line)
print(output)
```
cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @malfet @snadampal @milpuz01 @aditew01 @nikhil-arm @fadara01

### CODE REVIEW

**Comment by maajidkhann** [comment]
@swolchok Can you please add the label "ciflow/linux-aarch64" and trigger the pipelines.
And also review this PR whenever you get time.

File: aten/src/ATen/cpu/vec/sve/vec_float.h [hunk 1]
```diff
@@ -70,6 +70,57 @@ template <> class Vectorized<float> {
     }
     return b;
   }
+  inline svfloat32_t svexp_f32_z(svbool_t pg, svfloat32_t x) const {
+    const auto c1 = svreinterpret_f32_u32(svdup_n_u32(0x3f7ffff6)); // x^1: 0x1.ffffecp-1f
```
Comment (swolchok) [inline aten/src/ATen/cpu/vec/sve/vec_float.h:74]: can't speak to the acceptability of the numerics here at all, not my area of expertise. you'll need somebody who knows about approximations and PyTorch's expectations around them to accept.

Comment (maajidkhann) [reply aten/src/ATen/cpu/vec/sve/vec_float.h:74]: Thanks Scott.

Adding @malfet and @milpuz01 to this thread.

The changes in this PR have been picked from Arm Compute Library.
https://github.com/ARM-software/ComputeLibrary/blob/main/src/core/NEON/SVEMath.inl#L195 (tanh)
https://github.com/ARM-software/ComputeLibrary/blob/main/src/core/NEON/SVEMath.inl#L105 (exp)

ACL algorithm is the best compared to other algorithms like Eigen and others in terms of accuracy when comparing it with std. We have compared the accuracy of different algorithms and ACL seems to be the best w.r.t performance and accuracy.

Coming to this PR, The function getDefaultTolerance<UVT>() is used in the testcase which is likely a template which returns different tolerance values based on the type (UVT, short for Unit Vectorized Type) being tested. The tests pass successfully when tested on ARM CPU, so I believe this should be fine.

Lot other operators have been similarly adjusted during testing in the vec_test backend like ATan, Acos, LGamma, etc
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/test/vec_test_all_types.cpp#L423
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/test/vec_test_all_types.cpp#L438

Comment (malfet) [reply aten/src/ATen/cpu/vec/sve/vec_float.h:74]: Would be nice to embed above-mentioned references as permalinks into the source code

**Review by malfet — APPROVED** [review]
If it passes CI, than sure why not

Also, why not contribute similar change to sleef and just relying on the sleef?

File: aten/src/ATen/cpu/vec/sve/vec_float.h (hunk 1 above)
Comment (maajidkhann) [reply aten/src/ATen/cpu/vec/sve/vec_float.h:75]: Updated.

**Comment by maajidkhann** [comment]
> If it passes CI, than sure why not
> 
> Also, why not contribute similar change to sleef and just relying on the sleef?

This change taken from ACL is optimized for performance, sacrificing minor accuracy (within the threshold) in exchange for speed while the current tanh sve sleef implementation might probably give superior accuracy with reference to std.

Since this algorithm is well tailored just for ARM and gives great perf boost, I feel PyTorch Aten kernel might be the right place for this.

**Comment by maajidkhann** [comment]
> If it passes CI, than sure why not
> 
> Also, why not contribute similar change to sleef and just relying on the sleef?

Can you please re-trigger the CI pipelines again.

**Comment by maajidkhann** [comment]
@malfet Can you please merge this PR.

**Comment by maajidkhann** [comment]
> Pull workflow has not been scheduled for the PR yet. It could be because author doesn't have permissions to run those or skip-checks keywords were added to PR/commits, aborting merge. Please get/give approval for the workflows and/or remove skip ci decorators before next merge attempt. If you think this is a mistake, please contact PyTorch Dev Infra.

@malfet I think this needs your permissions to get merged. Can you please merge this!

**Comment by abhijain1204fujitsu** [comment]
@malfet , kindly support us to merge this PR looks like it might require your permission

**Comment by maajidkhann** [comment]
@malfet Can you please merge this when you get time as we are closing on cut off date for next PyTorch release.

**Comment by maajidkhann** [comment]
Hi @malfet . Can we merge this PR? Let me know if you have any concerns.

**Comment by maajidkhann** [comment]
> @maajidkhann can you please check if 3 ci failures are due to this change?

I think these failures are unrelated to my PR. Mu branch was old, I have rebased it w.r.t latest main.
Can you re trigger aarch64 CI please?

**Comment by maajidkhann** [comment]
> > @maajidkhann can you please check if 3 ci failures are due to this change?
> 
> I think these failures are unrelated to my PR. My branch was old, I have rebased it w.r.t latest main. Can you re trigger aarch64 CI please?
